File: src/Features/CSharp/Organizing/Organizers/MemberDeclarationsOrganizer.Comparer.cs
// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

using System.Collections.Generic;
using System.Linq;
using Microsoft.CodeAnalysis.CSharp.Extensions;
using Microsoft.CodeAnalysis.CSharp.Symbols;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.CSharp.Utilities;
using Microsoft.CodeAnalysis.Shared.Extensions;

namespace Microsoft.CodeAnalysis.CSharp.Organizing.Organizers
{
    internal partial class MemberDeclarationsOrganizer
    {
        private class Comparer : IComparer<MemberDeclarationSyntax>
        {
            // TODO(cyrusn): Allow users to specify the ordering they want
            private enum OuterOrdering
            {
                Fields,
                EventFields,
                Constructors,
                Destructors,
                Properties,
                Events,
                Indexers,
                Operators,
                ConversionOperators,
                Methods,
                Types,
                Remaining
            }

            private enum InnerOrdering
            {
                StaticInstance,
                Accessibility,
                Name
            }

            private enum Accessibility
            {
                Public,
                Protected,
                ProtectedOrInternal,
                Internal,
                Private
            }

            public int Compare(MemberDeclarationSyntax x, MemberDeclarationSyntax y)
            {
                if (x == y)
                {
                    return 0;
                }

                var xOuterOrdering = GetOuterOrdering(x);
                var yOuterOrdering = GetOuterOrdering(y);

                var compare = xOuterOrdering - yOuterOrdering;
                if (compare != 0)
                {
                    return compare;
                }

                if (xOuterOrdering == OuterOrdering.Remaining)
                {
                    return 1;
                }
                else if (yOuterOrdering == OuterOrdering.Remaining)
                {
                    return -1;
                }

                if (xOuterOrdering == OuterOrdering.Fields || yOuterOrdering == OuterOrdering.Fields)
                {
                    // Fields with initializers can't be reordered relative to 
                    // themselves due to ordering issues.
                    var xHasInitializer = ((FieldDeclarationSyntax)x).Declaration.Variables.Any(v => v.Initializer != null);
                    var yHasInitializer = ((FieldDeclarationSyntax)y).Declaration.Variables.Any(v => v.Initializer != null);
                    if (xHasInitializer && yHasInitializer)
                    {
                        return 0;
                    }
                }

                var xIsStatic = x.GetModifiers().Any(t => t.Kind() == SyntaxKind.StaticKeyword);
                var yIsStatic = y.GetModifiers().Any(t => t.Kind() == SyntaxKind.StaticKeyword);

                if ((compare = Comparer<bool>.Default.Inverse().Compare(xIsStatic, yIsStatic)) != 0)
                {
                    return compare;
                }

                var xAccessibility = GetAccessibility(x);
                var yAccessibility = GetAccessibility(y);
                if ((compare = xAccessibility - yAccessibility) != 0)
                {
                    return compare;
                }

                var xName = ShouldCompareByName(x) ? x.GetNameToken() : default(SyntaxToken);
                var yName = ShouldCompareByName(y) ? y.GetNameToken() : default(SyntaxToken);

                if ((compare = TokenComparer.NormalInstance.Compare(xName, yName)) != 0)
                {
                    return compare;
                }

                // Their names were the same.  Order them by arity at this point.
                return x.GetArity() - y.GetArity();
            }

            private static Accessibility GetAccessibility(MemberDeclarationSyntax x)
            {
                var xModifiers = x.GetModifiers();

                if (xModifiers.Any(t => t.Kind() == SyntaxKind.PublicKeyword))
                {
                    return Accessibility.Public;
                }
                else if (xModifiers.Any(t => t.Kind() == SyntaxKind.ProtectedKeyword) && xModifiers.Any(t => t.Kind() == SyntaxKind.InternalKeyword))
                {
                    return Accessibility.ProtectedOrInternal;
                }
                else if (xModifiers.Any(t => t.Kind() == SyntaxKind.InternalKeyword))
                {
                    return Accessibility.Internal;
                }
                else if (xModifiers.Any(t => t.Kind() == SyntaxKind.ProtectedKeyword))
                {
                    return Accessibility.Protected;
                }
                else
                {
                    return Accessibility.Private;
                }
            }

            private static OuterOrdering GetOuterOrdering(MemberDeclarationSyntax x)
            {
                switch (x.Kind())
                {
                    case SyntaxKind.FieldDeclaration:
                        return OuterOrdering.Fields;
                    case SyntaxKind.EventFieldDeclaration:
                        return OuterOrdering.EventFields;
                    case SyntaxKind.ConstructorDeclaration:
                        return OuterOrdering.Constructors;
                    case SyntaxKind.DestructorDeclaration:
                        return OuterOrdering.Destructors;
                    case SyntaxKind.PropertyDeclaration:
                        return OuterOrdering.Properties;
                    case SyntaxKind.EventDeclaration:
                        return OuterOrdering.Events;
                    case SyntaxKind.IndexerDeclaration:
                        return OuterOrdering.Indexers;
                    case SyntaxKind.OperatorDeclaration:
                        return OuterOrdering.Operators;
                    case SyntaxKind.ConversionOperatorDeclaration:
                        return OuterOrdering.ConversionOperators;
                    case SyntaxKind.MethodDeclaration:
                        return OuterOrdering.Methods;
                    case SyntaxKind.ClassDeclaration:
                    case SyntaxKind.InterfaceDeclaration:
                    case SyntaxKind.StructDeclaration:
                    case SyntaxKind.EnumDeclaration:
                    case SyntaxKind.DelegateDeclaration:
                        return OuterOrdering.Types;
                    default:
                        return OuterOrdering.Remaining;
                }
            }

            private static bool ShouldCompareByName(MemberDeclarationSyntax x)
            {
                // Constructors, destructors, indexers and operators should not be sorted by name.
                // Note:  Conversion operators should not be sorted by name either, but it's not
                //        necessary to deal with that here, because GetNameToken cannot return a
                //        name for them (there's only a NameSyntax, not a Token).
                switch (x.Kind())
                {
                    case SyntaxKind.ConstructorDeclaration:
                    case SyntaxKind.DestructorDeclaration:
                    case SyntaxKind.IndexerDeclaration:
                    case SyntaxKind.OperatorDeclaration:
                        return false;
                    default:
                        return true;
                }
            }
        }
    }
}

File: src/Features/VisualBasic/Organizing/Organizers/MemberDeclarationsOrganizer.Comparer.vb
' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic.Symbols
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
Imports Microsoft.CodeAnalysis.VisualBasic.Utilities

Namespace Microsoft.CodeAnalysis.VisualBasic.Organizing.Organizers
    Friend Partial Class MemberDeclarationsOrganizer
        Public Class Comparer
            Implements IComparer(Of StatementSyntax)
            ' TODO(cyrusn): Allow users to specify the ordering they want
            Public Enum OuterOrdering
                Fields
                EventFields
                Constructors
                Destructors
                Properties
                Events
                Indexers
                Operators
                ConversionOperators
                Methods
                Types
                Remaining
            End Enum

            Public Enum InnerOrdering
                StaticInstance
                Accessibility
                Name
            End Enum

            Public Enum Accessibility
                [Public]
                [Protected]
                [Friend]
                [Private]
            End Enum

            Public Function Compare(x As StatementSyntax, y As StatementSyntax) As Integer Implements IComparer(Of StatementSyntax).Compare
                If x Is y Then
                    Return 0
                End If

                Dim xOuterOrdering = GetOuterOrdering(x)
                Dim yOuterOrdering = GetOuterOrdering(y)

                Dim value = xOuterOrdering - yOuterOrdering
                If value <> 0 Then
                    Return value
                End If

                If xOuterOrdering = OuterOrdering.Remaining Then
                    Return 1
                ElseIf yOuterOrdering = OuterOrdering.Remaining Then
                    Return -1
                End If

                If xOuterOrdering = OuterOrdering.Fields OrElse yOuterOrdering = OuterOrdering.Fields Then
                    ' Fields with initializers can't be reordered relative to 
                    ' themselves due to ordering issues.
                    Dim xHasInitializer = DirectCast(x, FieldDeclarationSyntax).Declarators.Any(Function(v) v.Initializer IsNot Nothing)
                    Dim yHasInitializer = DirectCast(y, FieldDeclarationSyntax).Declarators.Any(Function(v) v.Initializer IsNot Nothing)
                    If xHasInitializer AndAlso yHasInitializer Then
                        Return 0
                    End If
                End If

                Dim xIsShared = x.GetModifiers().Any(Function(t) t.Kind = SyntaxKind.SharedKeyword)
                Dim yIsShared = y.GetModifiers().Any(Function(t) t.Kind = SyntaxKind.SharedKeyword)

                value = Comparer(Of Boolean).Default.Inverse().Compare(xIsShared, yIsShared)
                If value <> 0 Then
                    Return value
                End If

                Dim xAccessibility = GetAccessibility(x)
                Dim yAccessibility = GetAccessibility(y)
                value = xAccessibility - yAccessibility
                If value <> 0 Then
                    Return value
                End If

                Dim xName = If(ShouldCompareByName(x), TryCast(x, DeclarationStatementSyntax).GetNameToken(), Nothing)
                Dim yName = If(ShouldCompareByName(x), TryCast(y, DeclarationStatementSyntax).GetNameToken(), Nothing)

                value = TokenComparer.NormalInstance.Compare(xName, yName)
                If value <> 0 Then
                    Return value
                End If

                ' Their names were the same.  Order them by arity at this point.
                Return x.GetArity() - y.GetArity()
            End Function

            Private Shared Function GetAccessibility(x As StatementSyntax) As Accessibility
                Dim xModifiers = x.GetModifiers()

                If xModifiers.Any(Function(t) t.Kind = SyntaxKind.PublicKeyword) Then
                    Return Accessibility.Public
                ElseIf xModifiers.Any(Function(t) t.Kind = SyntaxKind.FriendKeyword) Then
                    Return Accessibility.Friend
                ElseIf xModifiers.Any(Function(t) t.Kind = SyntaxKind.ProtectedKeyword) Then
                    Return Accessibility.Protected
                Else
                    ' Only fields are private in VB. All other members are public
                    If x.Kind = SyntaxKind.FieldDeclaration Then
                        Return Accessibility.Private
                    Else
                        Return Accessibility.Public
                    End If
                End If
            End Function

            Private Function GetOuterOrdering(x As StatementSyntax) As OuterOrdering
                Select Case x.Kind
                    Case SyntaxKind.FieldDeclaration
                        Return OuterOrdering.Fields
                    Case SyntaxKind.ConstructorBlock
                        Return OuterOrdering.Constructors
                    Case SyntaxKind.PropertyBlock
                        Return OuterOrdering.Properties
                    Case SyntaxKind.EventBlock
                        Return OuterOrdering.Events
                    Case SyntaxKind.OperatorBlock
                        Return OuterOrdering.Operators
                    Case SyntaxKind.SubBlock,
                        SyntaxKind.FunctionBlock
                        Return OuterOrdering.Methods
                    Case SyntaxKind.ClassBlock,
                        SyntaxKind.InterfaceBlock,
                        SyntaxKind.StructureBlock,
                        SyntaxKind.EnumBlock,
                        SyntaxKind.DelegateSubStatement,
                        SyntaxKind.DelegateFunctionStatement
                        Return OuterOrdering.Types
                    Case Else
                        Return OuterOrdering.Remaining
                End Select
            End Function

            Private Shared Function ShouldCompareByName(x As StatementSyntax) As Boolean
                ' Constructors and operators should not be sorted by name.
                Select Case x.Kind
                    Case SyntaxKind.ConstructorBlock,
                         SyntaxKind.OperatorBlock
                        Return False
                    Case Else
                        Return True
                End Select
            End Function
        End Class
    End Class
End Namespace
